### ISSUE TITLE
Darwin doesn't support separateDebugInfo (which is also undocumented)

### ISSUE THREAD

**Issue by lf-** [NixOS/nixpkgs#203380]
### Describe the bug

`separateDebugInfo` is ignored on macOS, which seems to be undocumented. #51872 looks possibly related based on search.

Here is it being ignored:

https://github.com/nixos/nixpkgs/blob/341e6fd558abde5b4f6609833ff503c8d1e9b379/pkgs/stdenv/generic/make-derivation.nix#L165

### Steps To Reproduce
1. `nix-build '<nixpkgs>' -A python3.debug`
2. Attribute not found?!

### Expected behavior
I expect to have the separate debug symbols output available on Darwin.

### Notify maintainers
@matthewbauer perhaps

### Metadata

```console
eden/scm - [main●] » nix run nixpkgs#nix-info -- -m
 - system: `"aarch64-darwin"`
 - host os: `Darwin 21.6.0, macOS 12.6.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.11.0`
 - channels(root): `"nixpkgs"`
 - channels(jade): `""`
 - nixpkgs: `/nix/store/y63kal4f9mmdy14mdfp8d24h6hfi4cqp-source`
```

**Comment by mstone** [comment]
Thanks for reporting this, @lf-, this is indeed an unwelcome surprise.

Regarding the situation in the codebase: this reminds me of a somewhat related issue regarding support for working with debuginfo on macOS, #178366, that (happily!) has a draft PR that could use some testing and review love: #188347.

However, beyond that connection, I also just did some code spelunking to learn more about the history of `separateDebugInfo`'s platform support, from which I found:

* before `separateDebugInfo` came be ignored on darwin I think by d64f28815486924a6fc3b1d9a46e6925b5d3c2ba, there was an `isLinux` assert
* 2a28bc669124190e28f7d62663eee9ce31759bc8 is the commit that introduced the assert
* ISTM that the assert-introducing commit is the immediate successor on the relevant file [working backwards past a big refactoring commit that pulled a bunch of meta-checks out into a separate file) to ec5b66eb4add9d494d8fb16f6899028750d317a2 which is where the `separateDebugInfo` concept seems to have been originally introduced (see [history](https://github.com/NixOS/nixpkgs/commits/2a28bc669124190e28f7d62663eee9ce31759bc8/pkgs/stdenv/generic/default.nix) ). 

As a result, I wonder if maybe we're simply now in better place re being able to reasonably implement support for `separateDebugInfo` on darwin than was the case all these years ago?

**Comment by n8henrie** [comment]
For rust, one can specify `RUSTFLAGS='-C split-debuginfo=packed'` and the debug information is then put into a `.dSYM` directory that can easily be included in `$out` and seems to work.

https://github.com/NixOS/nixpkgs/issues/262131

**Comment by symphorien** [comment]
note that copying to $out is not perfect as debug symbols are very heavy.

**Comment by n8henrie** [comment]
@symphorien agreed that this should only be done when compiling in debug mode, but when that is the case, then getting those (heavy) debug symbols is the whole point.
